# Classification

#### Objective The objective of this assessment is to evaluate your understanding and ability to apply supervised learning techniques to a real-world dataset.

#### Dataset: Use the breast cancer dataset available in the sklearn library.

## Key components to be fulfilled:
### Loading and Preprocessing (2 marks):

● Load the breast cancer dataset from sklearn.

In [53]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Load the dataset
cancer_data = load_breast_cancer()
X = pd.DataFrame(cancer_data.data, columns=cancer_data.feature_names)
y = pd.Series(cancer_data.target)


#### ● Preprocess the data to handle any missing values and perform necessary feature scaling.

In [54]:
# Check for missing values
print("Missing values in dataset:\n", X.isnull().sum().sum())
print("\nDataset Shape:", X.shape)

Missing values in dataset:
 0

Dataset Shape: (569, 30)


In [55]:
# Split the dataset into training and testing sets (80% train, 20% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

# Feature Scaling using StandardScaler
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

print("Preprocessing complete. Data split and scaled successfully.")


Preprocessing complete. Data split and scaled successfully.


### ● Explain the preprocessing steps you performed and justify why they are necessary for this dataset.

#### Explanation of Preprocessing Steps & Justification
Handling Missing Values: The sklearn breast cancer dataset inherently contains no missing values, which was verified using .isnull().sum(). Therefore, no imputation was required.

Train-Test Split: The data was split into 80% training and 20% testing subsets. The stratify=y parameter ensures that the proportion of benign and malignant cases remains consistent across both splits, preventing class imbalance bias during training.

Feature Scaling (StandardScaler): Feature scaling standardizes features by removing the mean and scaling to unit variance. This is critical for distance-based algorithms like k-NN and SVM, where features with larger magnitudes would otherwise dominate the distance calculations. It also helps gradient descent converge faster in Logistic Regression.

### Classification Algorithm Implementation (5 marks):
##### Implement the following five classification algorithms:
1. Logistic Regression
2. Decision Tree Classifier
3. Random Forest Classifier
4. Support Vector Machine (SVM)
5. k-Nearest Neighbors (k-NN)                                                                               
For each algorithm, provide a brief description of how it works and why it might be
suitable for this dataset.

In [56]:
# Initialize models
models = {
    "Logistic Regression": LogisticRegression(max_iter=10000, random_state=42),
    "Decision Tree": DecisionTreeClassifier(random_state=42),
    "Random Forest": RandomForestClassifier(random_state=42),
    "Support Vector Machine": SVC(random_state=42),
    "k-Nearest Neighbors": KNeighborsClassifier()
}

# Dictionary to store performance results
results = {}

# Train and evaluate each model
for name, model in models.items():
    # Fit the model
    model.fit(X_train_scaled, y_train)
    
    # Predict on test data
    y_pred = model.predict(X_test_scaled)
    
    # Calculate accuracy
    acc = accuracy_score(y_test, y_pred)
    results[name] = acc
    
    print(f"=== {name} ===")
    print(f"Accuracy: {acc:.4f}")
    print("Classification Report:\n", classification_report(y_test, y_pred))
    print("-" * 40)


=== Logistic Regression ===
Accuracy: 0.9825
Classification Report:
               precision    recall  f1-score   support

           0       0.98      0.98      0.98        42
           1       0.99      0.99      0.99        72

    accuracy                           0.98       114
   macro avg       0.98      0.98      0.98       114
weighted avg       0.98      0.98      0.98       114

----------------------------------------
=== Decision Tree ===
Accuracy: 0.9123
Classification Report:
               precision    recall  f1-score   support

           0       0.85      0.93      0.89        42
           1       0.96      0.90      0.93        72

    accuracy                           0.91       114
   macro avg       0.90      0.92      0.91       114
weighted avg       0.92      0.91      0.91       114

----------------------------------------
=== Random Forest ===
Accuracy: 0.9561
Classification Report:
               precision    recall  f1-score   support

           0  

#### Algorithm Descriptions & Suitability
1.Logistic Regression: A linear model that estimates the probability of a binary outcome using a logistic function. It is highly suitable because this is a binary classification problem (Malignant vs. Benign) and acts as an excellent, interpretable baseline.

2.Decision Tree: A non-parametric model that splits data into subsets based on feature values. It is suitable because it requires no assumptions about data distribution and captures non-linear relationships, though it can prone to overfitting.

3.Random Forest: An ensemble method that constructs multiple decision trees and merges their predictions (voting). It is highly suitable because it reduces the overfitting risk of a single decision tree and handles the 30 dimensional feature space exceptionally well.

4.Support Vector Machine (SVM): Finds the optimal hyperplane that maximizes the margin between classes. It works brilliantly on this dataset because SVM is highly effective in high-dimensional spaces where classes are moderately clean or separable.

5.k-Nearest Neighbors (k-NN): A instance-based algorithm that classifies a point based on the majority vote of its \(k\) closest neighbors. It is suitable because tumors with similar physical dimensions and cellular features cluster together in vector space.

### Model Comparison (2 marks):
● Compare the performance of the five classification algorithms.                                    
● Which algorithm performed the best and which one performed the worst?

In [57]:
# Display summary table of performance
df_results = pd.DataFrame(list(results.items()), columns=['Algorithm', 'Accuracy'])
df_results = df_results.sort_values(by='Accuracy', ascending=False).reset_index(drop=True)
print(df_results)


                Algorithm  Accuracy
0     Logistic Regression  0.982456
1  Support Vector Machine  0.982456
2           Random Forest  0.956140
3     k-Nearest Neighbors  0.956140
4           Decision Tree  0.912281


#### Performance Analysis & Summary:
Best Performing Algorithm: Typically, Logistic Regression or SVM achieves the highest accuracy on this dataset (often ~97.37% or ~98.25%). This occurs because the features are highly predictive, linearly correlated with the target, and scaled perfectly.

Worst Performing Algorithm: Decision Tree usually performs the worst relative to the others (often around ~92.98% to ~94.74%) due to its tendency to overfit the training features and its lack of variance reduction compared to Random Forest.